### PR TITLE
Ensure that SymCrypt structures have sufficient alignment

### PR DESCRIPTION
**Pull request by samuel-lee-msft**
+ Previously SCOSSL would rely on allocations made by libcrypto being at least 16-byte aligned for 64-bit platforms. This was normally true, but in some environments where the allocator is modified, only 8-byte alignment was being provided.
+ Should address #59

### CODE REVIEW

File: ScosslCommon/inc/scossl_helpers.h [hunk 1]
```diff
@@ -36,6 +36,36 @@ typedef _Return_type_success_(return >= 0) int SCOSSL_RETURNLENGTH; // For funct
...
+    PBYTE aligned_alloc = NULL;                                 \
+    if (alloc)                                                  \
+    {                                                           \
+        aligned_alloc = SCOSSL_ALIGN_UP(alloc+1);               \
+        *(aligned_alloc - 1) = aligned_alloc - alloc;           \
+    }                                                           \
+    typename *ptr = (typename *) aligned_alloc;
+
+#define SCOSSL_COMMON_ALIGNED_FREE(ptr, deallocator, typename)  \
+    PBYTE aligned_alloc = (PBYTE) ptr;                          \
```
Comment (mlindgren) [inline ScosslCommon/inc/scossl_helpers.h:65]: >aligned_alloc [](http://example.com/codeflow?start=10&length=13)

Probably not a problem given the way it's used, but this shadows the standard library [aligned_alloc](https://en.cppreference.com/w/c/memory/aligned_alloc).

File: SymCryptProvider/src/p_scossl_base.c [hunk 2]
```diff
@@ -249,22 +249,7 @@ SCOSSL_STATUS OSSL_provider_init(_In_ const OSSL_CORE_HANDLE *handle,
                                  _Out_ const OSSL_DISPATCH **out,
                                  _Out_ void **provctx)
 {
-    SCOSSL_PROVCTX *p_ctx = OPENSSL_malloc(sizeof(SCOSSL_PROVCTX));
-    if (p_ctx != NULL)
-    {
-        p_ctx->handle = handle;
-        *provctx = p_ctx;
-    }
```
Comment (mlindgren) [inline SymCryptProvider/src/p_scossl_base.c:257]: Just curious, why was this moved?

**Review by mlindgren — DISMISSED** [review]
:shipit:

File: SymCryptProvider/src/p_scossl_base.c (hunk 2 above)
Comment (samuel-lee-msft) [reply SymCryptProvider/src/p_scossl_base.c:257]: The intent is to use the malloc function we are provided in the `in` dispatch table to allocate the provider context; rather than maybe accidentally linking libcrypto for the call to OPENSSL_malloc.

Comment (samuel-lee-msft) [reply SymCryptProvider/src/p_scossl_base.c:257]: @mamckee does this change make sense to you?

File: ScosslCommon/inc/scossl_helpers.h (hunk 1 above)
Comment (samuel-lee-msft) [reply ScosslCommon/inc/scossl_helpers.h:65]: Hrmm, that's a fair point - I'll rename to avoid ambiguity.

File: SymCryptProvider/src/ciphers/p_scossl_aes.c [hunk 3]
```diff
@@ -541,7 +541,7 @@ static SCOSSL_STATUS scossl_aes_cfb8_cipher(_Inout_ SCOSSL_AES_CTX *ctx,
 #define IMPLEMENT_SCOSSL_AES_BLOCK_CIPHER(kbits, ivlen, lcmode, UCMODE)                                   \
     SCOSSL_AES_CTX *p_scossl_aes_##kbits##_##lcmode##_newctx()                                            \
     {                                                                                                     \
-        SCOSSL_AES_CTX *ctx = OPENSSL_zalloc(sizeof(SCOSSL_AES_CTX));                                     \
+        SCOSSL_COMMON_ALIGNED_ALLOC(ctx, OPENSSL_malloc, SCOSSL_AES_CTX);                                 \
```
Comment (ccalik) [inline SymCryptProvider/src/ciphers/p_scossl_aes.c:544]: OPENSSL_zalloc

File: SymCryptProvider/src/ciphers/p_scossl_aes_xts.c [hunk 4]
```diff
@@ -38,8 +38,8 @@ static SCOSSL_STATUS p_scossl_aes_xts_set_ctx_params(_Inout_ SCOSSL_AES_XTS_CTX
 
 static SCOSSL_AES_XTS_CTX *p_scossl_aes_xtx_newctx_internal(size_t keylen)
 {
-    SCOSSL_AES_XTS_CTX *ctx = OPENSSL_zalloc(sizeof(SCOSSL_AES_XTS_CTX));
-    if (ctx != NULL)
+    SCOSSL_COMMON_ALIGNED_ALLOC(ctx, OPENSSL_malloc, SCOSSL_AES_XTS_CTX);
```
Comment (ccalik) [inline SymCryptProvider/src/ciphers/p_scossl_aes_xts.c:41]: OPENSSL_zalloc

File: SymCryptProvider/src/ciphers/p_scossl_aes_xts.c [hunk 5]
```diff
@@ -48,7 +48,7 @@ static SCOSSL_AES_XTS_CTX *p_scossl_aes_xtx_newctx_internal(size_t keylen)
 
 static SCOSSL_AES_XTS_CTX *p_scossl_aes_xts_dupctx(SCOSSL_AES_XTS_CTX *ctx)
 {
-    SCOSSL_AES_XTS_CTX *copy_ctx = OPENSSL_zalloc(sizeof(SCOSSL_AES_XTS_CTX));
+    SCOSSL_COMMON_ALIGNED_ALLOC(copy_ctx, OPENSSL_malloc, SCOSSL_AES_XTS_CTX);
```
Comment (ccalik) [inline SymCryptProvider/src/ciphers/p_scossl_aes_xts.c:51]: OPENSSL_zalloc

**Review by ccalik — DISMISSED** [review]

File: SymCryptProvider/src/ciphers/p_scossl_aes.c (hunk 3 above)
Comment (samuel-lee-msft) [reply SymCryptProvider/src/ciphers/p_scossl_aes.c:544]: Ah - I had intentionally made the change to be more consistent with AEAD versions; there's no security reason we need to zero before use and it seemed to be a waste of cycles.

However, we don't initialize cbBuf without the zalloc. We could explicitly initialize cbBuf to 0, or just zero the whole ctx as before.
Probably it's better to use zalloc in all newctx functions and malloc in the dupctx just to be consistent.

Comment (samuel-lee-msft) [reply SymCryptProvider/src/ciphers/p_scossl_aes.c:545]: Hmm, when I went to make the change I decided it was more consistent to just ensure that everything that needs to be initialized is initialized. I believe only cbBuf was not being set.

File: SymCryptProvider/src/ciphers/p_scossl_aes_xts.c (hunk 4 above)
Comment (samuel-lee-msft) [reply SymCryptProvider/src/ciphers/p_scossl_aes_xts.c:41]: Intentionally moving to malloc for consistency - there is no need for zero-ing

File: SymCryptProvider/src/ciphers/p_scossl_aes_xts.c (hunk 5 above)
Comment (samuel-lee-msft) [reply SymCryptProvider/src/ciphers/p_scossl_aes_xts.c:51]: Intentionally moving to malloc for consistency - there is no need for zero-ing

File: ScosslCommon/inc/scossl_helpers.h (hunk 1 above)
Comment (samuel-lee-msft) [reply ScosslCommon/inc/scossl_helpers.h]: Fixed

File: SymCryptProvider/src/p_scossl_base.c (hunk 2 above)
Comment (mamckee) [reply SymCryptProvider/src/p_scossl_base.c:257]: Yes, we don't need the provider handle to map these functions, so moving the function table fetch to the top makes sense.

File: SymCryptProvider/src/ciphers/p_scossl_aes.c (hunk 3 above)
Comment (mamckee) [reply SymCryptProvider/src/ciphers/p_scossl_aes.c:545]: It gets explicitly set in encrypt/decrypt init, but we should also zero it here. p_scossl_aes_generic_update assumes this value is between 0 and 15. The caller should call encrypt/decrypt init but if they don't, we might get an access violation rather than a simple error from p_scossl_aes_generic_update.

**Review by mamckee — DISMISSED** [review]

**Review by mlindgren — APPROVED** [review]
:shipit:

**Review by ccalik — APPROVED** [review]
